### PR TITLE
Add log system based on log4cplus

### PR DESCRIPTION
**Pull request by hongjingmao**
Add log system based on log4cplus

Signed-off-by: hongjingmao <hongjing.mao@intel.com>

### CODE REVIEW

**Comment by yang8621** [comment]
how to install log4cplus? add it to https://github.com/intel/ehsm/blob/main/docker/Dockerfile

File: utils/log4cplus/log4cplus_d.conf [hunk 1]
```diff
@@ -0,0 +1,10 @@
+log4cplus.rootLogger=TRACE,STDOUT,ALL_MSGS
```
Comment (yang8621) [inline utils/log4cplus/log4cplus_d.conf:1]: What's the difference between TRACE and INFO?

File: utils/log4cplus/log_utils.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#include <log4cplus/logger.h>
+#include <log4cplus/loggingmacros.h>
+#define IS_DEBUG true
+
+using namespace log4cplus;
+using namespace log4cplus::helpers;
+
+#if defined(__linux__)
+    #define VSPRINTF vsnprintf
+#elif defined(_WIN32)
```
Comment (yang8621) [inline utils/log4cplus/log_utils.h:44]: windows can be removed, we do not support it.

File: utils/log4cplus/log_utils.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+void log_d(const char* pFormat,  ...);
+void log_i(const char* pFormat,  ...);
+void log_e(const char* pFormat,  ...);
+void log_w(const char* pFormat,  ...);
+
+void initLogger(bool isDebug);
+ 
+void shutDown();
+ 
+void close();
```
Comment (yang8621) [inline utils/log4cplus/log_utils.h:61]: anywhere close() and shutDown() are called?
And I didn't find the definition of close().

File: dkeycache/Enclave/enclave.edl [hunk 4]
```diff
@@ -49,7 +49,7 @@ enclave {
     include "poll.h"
 
     untrusted {
-        void ocall_print_string([in, string] const char *str);
+        void ocall_print_string(uint32 log_type, [in, string] const char *str);
```
Comment (yang8621) [inline dkeycache/Enclave/enclave.edl:52]: uint32_t

File: enroll_app/main.cpp [hunk 5]
```diff
@@ -43,6 +43,7 @@ using namespace std;
 
 int main(int argc, char *argv[])
 {
+    initLogger(IS_DEBUG);
```
[the remark this answers is not in the document]
Comment (hongjingmao) [reply enroll_app/main.cpp:46]: done.

File: dkeycache/Enclave/enclave.edl (hunk 4 above)
Comment (hongjingmao) [reply dkeycache/Enclave/enclave.edl:52]: done.

File: utils/log4cplus/log_utils.h (hunk 3 above)
Comment (hongjingmao) [reply utils/log4cplus/log_utils.h:61]: done.

File: utils/log4cplus/log_utils.h (hunk 2 above)
Comment (hongjingmao) [reply utils/log4cplus/log_utils.h:44]: done.

File: utils/log4cplus/log4cplus_d.conf (hunk 1 above)
Comment (hongjingmao) [reply utils/log4cplus/log4cplus_d.conf:1]: They represent different levels of log output. Priority from low to high includes TRACE ->DEBUG ->INFO ->WARN

File: utils/log4cplus/log_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
+#include "log_utils.h"
+ 
+#include <log4cplus/logger.h>
+#include <log4cplus/layout.h>
+#include <log4cplus/configurator.h>
+ 
+Logger logger = Logger::getInstance(LOG4CPLUS_TEXT("logmain"));
+ 
+void initLogger(bool isDebug) {
```
Comment (wanghouqi) [inline utils/log4cplus/log_utils.cpp:9]: Here you need to add a parameter [prefix] as the prefix of the output log.

File: utils/log4cplus/log_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    va_start(args, pFormat);        \
+    char buf[bufSize] = {0};        \
+    VSPRINTF(buf, sizeof(buf), pFormat, args);    \
+    va_end(args);           \
+    logger.forcedLog(logLevel, buf); \
+}
+
+void log_d(const char* pFormat,  ...)
+{
+    DO_LOGGER(log4cplus::DEBUG_LOG_LEVEL, pFormat, 1024);
```
Comment (wanghouqi) [inline utils/log4cplus/log_utils.cpp:30]: Here need to print more information.
eg.  prfix  datetime  __FILE__  __LINE__  __FUNCTION__  ::  pformat

Comment (wanghouqi) [reply utils/log4cplus/log_utils.cpp:30]: ... Didn't seem to handle it.

File: utils/log4cplus/log_utils.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+{
+    DO_LOGGER(log4cplus::INFO_LOG_LEVEL, pFormat, 1024);
+}
+
+void log_w(const char* pFormat,  ...)
+{
+    DO_LOGGER(log4cplus::WARN_LOG_LEVEL, pFormat, 1024);
+}
+
+void shutDown(){
```
Comment (yang8621) [inline utils/log4cplus/log_utils.cpp:48]: rename to loggerShutdown or something else.

File: buildenv.mk [hunk 9]
```diff
@@ -73,6 +73,7 @@ RM = rm -f
 OPENSSL_PATH := $(ROOT_DIR)/utils/sgxssl
 OPENSSL_LIBRARY_PATH := $(OPENSSL_PATH)/lib64
 SOCKET_DIR := $(ROOT_DIR)/utils/sgx_socket
+LOG_DIR := $(ROOT_DIR)/utils/log4cplus
```
Comment (yang8621) [inline buildenv.mk:76]: not used?

File: utils/log4cplus/Makefile [hunk 10]
```diff
@@ -0,0 +1,6 @@
+all: log_utils.o
```
Comment (yang8621) [inline utils/log4cplus/Makefile:1]: a little strange to build .o here.

**Comment by yang8621** [comment]
remove this from the patch: utils/log4cplus/log_utils.o

File: core/App/ehsm_provider.cpp [hunk 11]
```diff
@@ -285,6 +309,7 @@ uint32_t EHSM_FFI_CALL(const char *reqJson, char *respJson)
 
 ehsm_status_t Initialize()
 {
+    initLogger(IS_DEBUG);
```
Comment (syan10) [inline core/App/ehsm_provider.cpp:312]: IS_DEBUG 啥意思

File: include/log_utils_e.h [hunk 12]
```diff
@@ -29,8 +29,8 @@
  *
  */
 
-#ifndef _LOG_UTILS_H
```
Comment (syan10) [inline include/log_utils_e.h:32]: 为啥改成log_utils_e.h?

File: include/log_utils_e.h [hunk 13]
```diff
@@ -40,14 +40,18 @@
...
+} log_type;
 /*
     print info
 */
 #define log_i(format, args...)                                             \
     {                                                                      \
-        printf("INFO [%s(%d) -> %s]: ", __FILE__, __LINE__, __FUNCTION__); \
-        printf(format, ##args);                                            \
-        printf("\n");                                                      \
+        ocall_print_string(LOG_INFO, format, args...);                         \
```
Comment (syan10) [inline include/log_utils_e.h:54]: 现在里面， 外面都会走到ocall_?

File: utils/log4cplus/log_utils.h [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+extern Logger logger;
+
+// define some macros for simplicity
+void log_d(const char* pFormat,  ...);
+void log_i(const char* pFormat,  ...);
+void log_e(const char* pFormat,  ...);
+void log_w(const char* pFormat,  ...);
+
+void initLogger(bool isDebug);
+
```
Comment (syan10) [inline utils/log4cplus/log_utils.h:54]: 把log_utils.o加到.ignore里面

File: utils/log4cplus/log_utils.h [hunk 15]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2011-2020 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Comment (syan10) [inline utils/log4cplus/log_utils.h:4]: log_utils.h, log_utils_e.h 啥区别

File: utils/log4cplus/log4cplus.conf [hunk 16]
```diff
@@ -0,0 +1,10 @@
+log4cplus.rootLogger=INFO,STDOUT,ALL_MSGS
+log4cplus.appender.STDOUT=log4cplus::ConsoleAppender
+log4cplus.appender.STDOUT.layout=log4cplus::PatternLayout
```
Comment (syan10) [inline utils/log4cplus/log4cplus.conf:3]: conf文件是怎么用的？

File: utils/log4cplus/Makefile (hunk 10 above)
Comment (syan10) [reply utils/log4cplus/Makefile:1]: 这个有必要单独make吗， 我觉得可以直接跟其他service 一起source code build？

Comment (yang8621) [reply utils/log4cplus/Makefile:1]: 一般要么自己有Makefile， build出一个.a
要么自己没有makefile，跟别的一起build成.o

**Comment by yang8621** [comment]
> how to install log4cplus? add it to https://github.com/intel/ehsm/blob/main/docker/Dockerfile

not update?

**Comment by yang8621** [comment]
I cannot make pass:

App/ehsm_provider.cpp: In function ‘ehsm_status_t Initialize()’:
App/ehsm_provider.cpp:312:24: error: too few arguments to function ‘void initLogger(bool, const char*)’
  312 |     initLogger(IS_DEBUG);
      |                        ^
In file included from App/ehsm_provider.cpp:54:
/home/ehsm/github/ehsm/utils/log4cplus/log_utils.h:53:6: note: declared here
   53 | void initLogger(bool isDebug, const char* properties_filename);
      |      ^~~~~~~~~~
make[1]: *** [Makefile:238: App/ehsm_provider.o] Error 1
make[1]: Leaving directory '/home/ehsm/github/ehsm/core'
make: *** [Makefile:41: all] Error 1

File: buildenv.mk (hunk 9 above)
Comment (hongjingmao) [reply buildenv.mk:76]: done.

File: utils/log4cplus/log_utils.cpp (hunk 8 above)
Comment (hongjingmao) [reply utils/log4cplus/log_utils.cpp]: done.

File: utils/log4cplus/Makefile (hunk 10 above)
Comment (hongjingmao) [reply utils/log4cplus/Makefile]: done.

File: core/App/ehsm_provider.cpp (hunk 11 above)
Comment (syan10) [reply core/App/ehsm_provider.cpp:312]: remove

File: dkeycache/Makefile [hunk 17]
```diff
@@ -193,6 +197,10 @@ Enclave/%.o: Enclave/%.cpp Enclave/enclave_t.h
 	@$(CXX) $(SGX_COMMON_CXXFLAGS) $(Enclave_Cpp_Flags) -c $< -o $@
 	@echo "CXX  <=  $<"
 
+../utils/sgx_socket/tls/%.o: ../utils/sgx_socket/tls/%.cpp Enclave/enclave_t.h
+	@$(CXX) $(SGX_COMMON_CXXFLAGS) $(Enclave_Cpp_Flags) -c $< -o $@
+	@echo "CXX  <=  $<"
```
Comment (syan10) [inline dkeycache/Makefile:202]: check

File: dkeycache/Makefile [hunk 18]
```diff
@@ -193,6 +197,10 @@ Enclave/%.o: Enclave/%.cpp Enclave/enclave_t.h
...
+
```
Comment (yang8621) [inline dkeycache/Makefile:203]: required?

File: include/log_utils_e.h [hunk 19]
```diff
@@ -40,44 +40,43 @@
...
-        printf("\n");                                                      \
+        log_printf(LOG_INFO, ##__VA_ARGS__);                         \
     }
 /*
     print debug
 */
-#define log_d(format, args...)                                                  \
+#define log_d(...)                                                  \
     {                                                                           \
         if (IS_DEBUG)                                                           \
```
Comment (syan10) [inline include/log_utils_e.h:62]: remove the IS_DEBUG

File: core/Enclave/key_operation.cpp [hunk 20]
```diff
@@ -37,14 +37,25 @@
...
+}
+
+void log_printf(uint32_t log_type, const char *fmt, ...)
+{
+    char buf[BUFSIZ] = {'\0'};
+    va_list ap;
+    va_start(ap, fmt);
+    vsnprintf(buf, BUFSIZ, fmt, ap);
+    va_end(ap);
+    ocall_print_string(log_type, buf);
```
Comment (yang8621) [inline core/Enclave/key_operation.cpp:58]: log_printf and  printf, keep only one.

File: core/Makefile [hunk 21]
```diff
@@ -54,9 +54,11 @@ Signed_Enclave_FileName := $(OUT)/$(Signed_Enclave_Name)
 ######## App Settings ########
 App_Cpp_Files := $(wildcard App/*.cpp)
 App_Cpp_Files += $(wildcard Test/*.cpp)
+App_Cpp_Files += $(wildcard ../utils/log4cplus/*.cpp)
```
Comment (yang8621) [inline core/Makefile:57]: can use $(LOG_DIR) here ?

File: include/log_utils_e.h (hunk 12 above)
Comment (yang8621) [reply include/log_utils_e.h:32]: rename to elog_utils.h

**Comment by yang8621** [comment]
还是build不过

File: include/log_utils_e.h (hunk 12 above)
Comment (syan10) [reply include/log_utils_e.h:32]: 把这个移到log4cplus folder下面去吧， 命名的话统一一下：
log_utils_u.h   -> 外边使用 
log_utils_t.h    -> enclave里面使用

**Comment by yang8621** [comment]
这个别忘了加:
how to install log4cplus? add it to https://github.com/intel/ehsm/blob/main/docker/Dockerfile

File: utils/log4cplus/log_utils.cpp (hunk 7 above)
Comment (hongjingmao) [reply utils/log4cplus/log_utils.cpp:46]: done.

File: core/App/ehsm_provider.cpp (hunk 11 above)
Comment (hongjingmao) [reply core/App/ehsm_provider.cpp]: done.

File: utils/log4cplus/log_utils.h (hunk 14 above)
Comment (hongjingmao) [reply utils/log4cplus/log_utils.h:54]: done.

File: dkeycache/Makefile (hunk 18 above)
Comment (hongjingmao) [reply dkeycache/Makefile]: done.

File: dkeyserver/Enclave/enclave.cpp [hunk 22]
```diff
@@ -62,14 +62,25 @@ typedef struct SocketMsgHandlerParam
...
 void printf(const char *fmt, ...)
 {
     char buf[BUFSIZ] = {'\0'};
     va_list ap;
     va_start(ap, fmt);
     vsnprintf(buf, BUFSIZ, fmt, ap);
     va_end(ap);
-    ocall_print_string(buf);
+    ocall_print_string(LOG_INFO, buf);
+}
```
[the remark this answers is not in the document]
Comment (hongjingmao) [reply dkeyserver/Enclave/enclave.cpp]: done.

File: core/Enclave/key_operation.cpp (hunk 20 above)
Comment (hongjingmao) [reply core/Enclave/key_operation.cpp:47]: done.

File: core/Makefile (hunk 21 above)
Comment (hongjingmao) [reply core/Makefile]: done.

File: include/log_utils_e.h (hunk 12 above)
Comment (hongjingmao) [reply include/log_utils_e.h]: done.

File: core/Enclave/enclave_msg_exchange.cpp [hunk 23]
```diff
@@ -46,7 +47,7 @@
 #include "marshal.h"
 #include "enclave_msg_exchange.h"
 
-extern void printf(const char *fmt, ...);
+extern void log_printf(uint32_t log_type, const char *fmt, ...);
```
Comment (syan10) [inline core/Enclave/enclave_msg_exchange.cpp:50]: 这个应该不需要了把

File: core/App/untrusted_enclave_msg_exchange.cpp [hunk 24]
```diff
@@ -61,7 +62,7 @@ extern "C" ATTESTATION_STATUS ocall_session_request(sgx_dh_msg1_t* dh_msg1, uint
 	if ((client_send_receive(&msg1_request, sizeof(FIFO_MSG), &msg1_response, &msg1_resp_size) != 0)
 		|| (msg1_response == NULL))
 	{
-		printf("fail to send and receive message.\n");
+		log_i("fail to send and receive message.\n");
```
Comment (syan10) [inline core/App/untrusted_enclave_msg_exchange.cpp:65]: 如果是error打印的话， 应该改成log_e

File: dkeycache/App/main.cpp [hunk 25]
```diff
@@ -155,6 +178,7 @@ static void parse_args(int argc, char *argv[])
 
 int main(int argc, char *argv[])
 {
+    initLogger("../../out/log/dkeycache.log");
```
Comment (syan10) [inline dkeycache/App/main.cpp:181]: 把log存到默认/var/run/ehsm下面把， 格式类似这样：
2022-12-20-ehsm-dkeycache.log
2022-12-20-ehsm-dkeyserver.log
2022-12-20-ehsm-core.log

File: include/datatypes.h [hunk 26]
```diff
@@ -35,7 +35,7 @@
 #include "sgx_dh.h"
 #include "sgx_tseal.h"
 
-#include "log_utils.h"
+// #include "log_utils.h"
```
Comment (syan10) [inline include/datatypes.h:38]: 直接删掉就行了

File: core/Makefile [hunk 27]
```diff
@@ -230,6 +234,10 @@ $(Signed_Enclave_Name): $(Enclave_Name)
 	@echo "SIGN =>  $@"
 
 ######## FFI Objects ########
+$(LOG_DIR)/%.o: $(LOG_DIR)/%.cpp
+	@$(CXX) $(Provider_Cpp_Flags) -llog4cplus -c $< -o $@
```
Comment (yang8621) [inline core/Makefile:238]: -llog4cplus 可以加到 Provider_Link_Flags里吗？

File: docker/Dockerfile [hunk 28]
```diff
@@ -28,7 +28,9 @@ RUN apt-get update && apt-get install -y \
     module-init-tools \
     g++ \
     libjsoncpp-dev\
-    uuid-dev
+    uuid-dev\
+    liblog4cplus-1.1-9\
+    liblog4cplus-dev
```
Comment (yang8621) [inline docker/Dockerfile:33]: please update the file due to conflict

File: dkeycache/App/main.cpp (hunk 25 above)
Comment (yang8621) [reply dkeycache/App/main.cpp:181]: 可以用这个RUNTIME_FOLDER ，  makefile需要加一下, 在makefile里拼出文件名的绝对路径

可以参考这个patch: https://github.com/intel/ehsm/pull/221/files

File: utils/log4cplus/log_utils_t.h [hunk 29]
```diff
@@ -29,8 +29,8 @@
  *
  */
 
-#ifndef _LOG_UTILS_H
-#define _LOG_UTILS_H
+#ifndef _LOG_UTILS_T_H
+#define _LOG_UTILS_T_H
```
Comment (yang8621) [inline utils/log4cplus/log_utils_t.h:34]: 唯一有点奇怪的是目前_t   _u的所有file都是自动生成的
@syan10

File: core/App/untrusted_enclave_msg_exchange.cpp (hunk 24 above)
Comment (hongjingmao) [reply core/App/untrusted_enclave_msg_exchange.cpp]: done.

File: include/datatypes.h (hunk 26 above)
Comment (hongjingmao) [reply include/datatypes.h]: done.

**Comment by hongjingmao** [comment]
> 编译有一些warnings， 如果好fix就先改一下

auto_ptr的warning是log4cplus版本太老了，只能接受auto_ptr

File: enroll_app/Makefile [hunk 30]
```diff
@@ -72,7 +74,11 @@ clean:
 %.o: %.cpp
 	@$(CXX) $(App_Cpp_Flags) -c $< -o $@
 	@echo "CXX  <=  $<"
-	
+
+$(LOG_DIR)/%.o: $(LOG_DIR)/%.cpp
+	@$(CXX) $(App_Cpp_Flags) -llog4cplus -c $< -o $@
```
[the remark this answers is not in the document]
Comment (hongjingmao) [reply enroll_app/Makefile]: done

File: dkeyserver/Makefile [hunk 31]
```diff
@@ -157,19 +160,23 @@ App/auto_version.h:
...
-	@echo "CC   <=  $<"
+	@$(CXX) $(SGX_COMMON_CFLAGS) $(App_C_Flags) -c $< -o $@
+	@echo "CXX   <=  $<"
 
 App/%.o: App/%.cpp App/auto_version.h
 	@$(CXX) $(SGX_COMMON_CXXFLAGS) $(App_Cpp_Flags) -c $< -o $@
 	@echo "CXX  <=  $<"
 
+$(LOG_DIR)/%.o: $(LOG_DIR)/%.cpp
+	@$(CXX) $(App_Cpp_Flags) -llog4cplus -c $< -o $@
```
[the remark this answers is not in the document]
Comment (hongjingmao) [reply dkeyserver/Makefile]: done.

File: dkeycache/Makefile [hunk 32]
```diff
@@ -156,19 +157,23 @@ App/auto_version.h:
...
-	@echo "CC   <=  $<"
+	@$(CXX) $(SGX_COMMON_CFLAGS) $(App_C_Flags) -c $< -o $@
+	@echo "CXX   <=  $<"
 
 App/%.o: App/%.cpp App/auto_version.h
 	@$(CXX) $(SGX_COMMON_CXXFLAGS) $(App_Cpp_Flags) -c $< -o $@
 	@echo "CXX  <=  $<"
 
+$(LOG_DIR)/%.o: $(LOG_DIR)/%.cpp
+	@$(CXX) $(App_Cpp_Flags) -llog4cplus -c $< -o $@
```
[the remark this answers is not in the document]
Comment (hongjingmao) [reply dkeycache/Makefile]: done.

File: core/Makefile (hunk 27 above)
Comment (hongjingmao) [reply core/Makefile]: done.

**Comment by yang8621** [comment]
> > 编译有一些warnings， 如果好fix就先改一下
> 
> auto_ptr的warning是log4cplus版本太老了，只能接受auto_ptr

可以的

这个看一下   换个名字应该可以解决：
Enclave/enclave_ra.cpp: In function ‘void log_printf(uint32_t, const char*, ...)’:
Enclave/enclave_ra.cpp:53:56: warning: declaration of ‘log_type’ shadows a global declaration [-Wshadow]
   53 | void log_printf(uint32_t log_type, const char *fmt, ...)
      |                                                        ^
In file included from /home/ehsm/github/ehsm/utils/sgx_socket/tls/common.h:35,
                 from /home/ehsm/github/ehsm/utils/sgx_socket/tls/utility.h:34,
                 from /home/ehsm/github/ehsm/utils/sgx_socket/tls/openssl_utility.h:30,
                 from Enclave/enclave_ra.cpp:44:
/home/ehsm/github/ehsm/utils/log4cplus/log_utils_t.h:48:3: note: shadowed declaration is here
   48 | } log_type;
      |   ^~~~~~~~

File: dkeycache/App/main.cpp (hunk 25 above)
Comment (hongjingmao) [reply dkeycache/App/main.cpp]: RUNTIME_FOLDER已更新

**Comment by hongjingmao** [comment]
> > > 编译有一些warnings， 如果好fix就先改一下
> > 
> > 
> > auto_ptr的warning是log4cplus版本太老了，只能接受auto_ptr
> 
> 可以的
> 
> 这个看一下 换个名字应该可以解决： Enclave/enclave_ra.cpp: In function ‘void log_printf(uint32_t, const char*, ...)’: Enclave/enclave_ra.cpp:53:56: warning: declaration of ‘log_type’ shadows a global declaration [-Wshadow] 53 | void log_printf(uint32_t log_type, const char *fmt, ...) | ^ In file included from /home/ehsm/github/ehsm/utils/sgx_socket/tls/common.h:35, from /home/ehsm/github/ehsm/utils/sgx_socket/tls/utility.h:34, from /home/ehsm/github/ehsm/utils/sgx_socket/tls/openssl_utility.h:30, from Enclave/enclave_ra.cpp:44: /home/ehsm/github/ehsm/utils/log4cplus/log_utils_t.h:48:3: note: shadowed declaration is here 48 | } log_type; | ^~~~~~~~

已更新

File: utils/log4cplus/log_utils.cpp [hunk 33]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    printf("Logs folder:\t%s", LOGS_FOLDER);
+    int path_len = strlen(LOGS_FOLDER) + strlen(logs_filename) + strlen("/");
+    char logs_path[path_len] = {0};        
+    sprintf(logs_path,"%s/%s", LOGS_FOLDER, logs_filename);
+
+    log4cplus::initialize ();
+
+    SharedAppenderPtr fileAppender(new RollingFileAppender(
```
Comment (syan10) [inline utils/log4cplus/log_utils.cpp:40]: 如果文件已经存在， 比如服务重启，  会发生什么

Comment (hongjingmao) [reply utils/log4cplus/log_utils.cpp:40]: 会在这个文件后面继续写

File: utils/log4cplus/elog_utils.h [hunk 34]
```diff
@@ -0,0 +1,67 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef _LOG_UTILS_T_H
+#define _LOG_UTILS_T_H
```
Comment (yang8621) [inline utils/log4cplus/elog_utils.h:33]: align with file name

File: enroll_app/main.cpp [hunk 35]
```diff
@@ -37,16 +37,22 @@ using namespace std;
...
 #include "rest_utils.h"
 #include "enroll_msg.h"
 
 int main(int argc, char *argv[])
 {
-    log_i("ehsm-kms enroll app start.");
     enroll_status_t ret = ENL_OK;
     RetJsonObj retJsonObj;
 
+    if (initLogger("enroll.log") < 0)
```
Comment (yang8621) [inline enroll_app/main.cpp:49]: 我在想这个app是否需要enable  log
加上了之后要root权限才能跑吧

File: enroll_app/main.cpp [hunk 36]
```diff
@@ -73,8 +79,8 @@ int main(int argc, char *argv[])
                 break;
             default:
 		if (ehsm_kms_url.empty()) {
-                    printf("\nusage: ehsm-kms_enroll_app -a https://1.2.3.4:9000/ehsm/ [-n]");
-                    printf("\n -n Optional only for informal certificates.\n\n");
+                    log_i("\nusage: ehsm-kms_enroll_app -a https://1.2.3.4:9000/ehsm/ [-n]");
+                    log_i("\n -n Optional only for informal certificates.\n\n");
```
Comment (yang8621) [inline enroll_app/main.cpp:83]: 这些是要打在屏幕上的   还会在有吗

Comment (hongjingmao) [reply enroll_app/main.cpp:83]: 有的，同时输出到console和文件

File: utils/log4cplus/elog_utils.h (hunk 34 above)
Comment (hongjingmao) [reply utils/log4cplus/elog_utils.h]: done.

File: enroll_app/main.cpp (hunk 35 above)
Comment (hongjingmao) [reply enroll_app/main.cpp:49]: 需要root权限才能打开log存储的文件夹

Comment (yang8621) [reply enroll_app/main.cpp:49]: enroll app我们就不改了吧   这个是给end user去跑的    还需要搞个root权限没必要

Comment (yang8621) [reply enroll_app/main.cpp:49]: enroll app还是用printf打log就行了

Comment (yang8621) [reply enroll_app/main.cpp:49]: 或者有没有办法可以不用root权限？  比如不写进file里？

Comment (hongjingmao) [reply enroll_app/main.cpp:49]: 单独enroll app不写进文件的话得用两套log4cplus的初始化函数

Comment (hongjingmao) [reply enroll_app/main.cpp]: done.

**Review by yang8621 — APPROVED** [review]

File: .gitignore [hunk 37]
```diff
@@ -20,6 +20,7 @@ utils/tkey_exchange/sgx_tkey_exchange_t.*
 utils/tkey_exchange/sgx_tkey_exchange_u.*
```
Comment (syan10) [inline .gitignore:20]: "
Add log system based on log4cplus

Store log files for each service separately in the runtime folder: /var/run/ehsm

The maximum capacity of a single log file is 10MB and each service can reserve
up to 5 log files. If the log number exceeds the maximum capacity, the oldest
log will be overwritten automatically.

test: passed all the unit test
Signed-off-by: hongjingmao <hongjing.mao@intel.com>
"

**Review by syan10 — APPROVED** [review]

**Pull request merged**